internal/controller/messaging/callback.go

```diff
@@ -67,42 +68,73 @@ func DeviceCallback(ctx context.Context, dic *di.Container) errors.EdgeX {
-					continue
-				}
-
-				var device dtos.Device
-				err = systemEvent.DecodeDetails(&device)
-				if err != nil {
-					lc.Errorf("failed to decode system event details: %s", err.Error())
-					continue
-				}
-
-				switch systemEvent.Action {
-				case common.DeviceSystemEventActionAdd:
-					err = application.AddDevice(requests.NewAddDeviceRequest(device), dic)
+				switch systemEvent.Type {
+				case common.DeviceSystemEventType:
+					err = deviceSystemEventAction(systemEvent, dic)
 					if err != nil {
 						lc.Error(err.Error(), common.CorrelationHeader, msgEnvelope.CorrelationID)
 					}
-				case common.DeviceSystemEventActionUpdate:
-					deviceModel := dtos.ToDeviceModel(device)
-					updateDeviceDTO := dtos.FromDeviceModelToUpdateDTO(deviceModel)
-					err = application.UpdateDevice(requests.NewUpdateDeviceRequest(updateDeviceDTO), dic)
-					if err != nil {
-						lc.Error(err.Error(), common.CorrelationHeader, msgEnvelope.CorrelationID)
-					}
-				case common.DeviceSystemEventActionDelete:
-					err = application.DeleteDevice(device.Name, dic)
+				case common.DeviceProfileSystemEventType:
+					err = deviceProfileSystemEventAction(systemEvent, dic)
 					if err != nil {
 						lc.Error(err.Error(), common.CorrelationHeader, msgEnvelope.CorrelationID)
 					}
 				default:
-					lc.Errorf("unknown device system event action %s", systemEvent.Action)
+					lc.Errorf("unknown system event type %s", systemEvent.Type)
+					continue
 				}
 			}
 		}
 	}()
 
 	return nil
 }
+
+func deviceSystemEventAction(systemEvent dtos.SystemEvent, dic *di.Container) error {
+	var device dtos.Device
+	err := systemEvent.DecodeDetails(&device)
+	if err != nil {
+		return fmt.Errorf("failed to decode %s system event details: %s", systemEvent.Type, err.Error())
+	}
+
+	switch systemEvent.Action {
+	case common.SystemEventActionAdd:
+		err = application.AddDevice(requests.NewAddDeviceRequest(device), dic)
+	case common.SystemEventActionUpdate:
+		deviceModel := dtos.ToDeviceModel(device)
+		updateDeviceDTO := dtos.FromDeviceModelToUpdateDTO(deviceModel)
+		err = application.UpdateDevice(requests.NewUpdateDeviceRequest(updateDeviceDTO), dic)
+	case common.SystemEventActionDelete:
+		err = application.DeleteDevice(device.Name, dic)
+	default:
+		return fmt.Errorf("unknown %s system event action %s", systemEvent.Type, systemEvent.Action)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deviceProfileSystemEventAction(systemEvent dtos.SystemEvent, dic *di.Container) error {
+	var deviceProfile dtos.DeviceProfile
+	err := systemEvent.DecodeDetails(&deviceProfile)
+	if err != nil {
+		return fmt.Errorf("failed to decode %s system event details: %s", systemEvent.Type, err.Error())
+	}
+
+	switch systemEvent.Action {
+	case common.SystemEventActionUpdate:
+		err = application.UpdateProfile(requests.NewDeviceProfileRequest(deviceProfile), dic)
+	case common.SystemEventActionAdd, common.SystemEventActionDelete:
+		break
```

Comment:
Why is no action taken for Add or Delete? Covered elsewhere?

Reply:
There are no previously callbacks for Add and Delete, because no associated device in cache exists at that point.

Reply:
add a comment `// there is no action needed for Device Profile Add and Delete in Device Service`

Reply:
Added the comment.